File: cg/resolver.py
import rclpy
from rclpy.node import Node
from cg_interfaces.srv import MoveCmd, GetMap, Reset
from collections import deque
import time

class MazeResolver(Node):
    def __init__(self):
        super().__init__('maze_resolver')
        
        # Clientes pros servicos dos comandos
        self.move_client = self.create_client(MoveCmd, '/move_command')
        self.map_client = self.create_client(GetMap, '/get_map')
        self.reset_client = self.create_client(Reset, '/reset')

        # Esperar serviços (delay)
        while not self.move_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Aguardando /move_command...')
        while not self.map_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Aguardando /get_map...')
        while not self.reset_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Aguardando /reset...')

    def reset_game(self):
        """Reinicia o jogo no mapa atual"""
        request = Reset.Request()
        request.is_random = True  
        request.map_name = ''      
        
        future = self.reset_client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        return future.result()

    def get_map_data(self):
        request = GetMap.Request()
        future = self.map_client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        return future.result()

    def move_robot(self, direction):
        request = MoveCmd.Request()
        request.direction = direction
        future = self.move_client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        return future.result()

    def bfs(self, grid, start, target, rows, cols):
        queue = deque([(start, [])])
        visited = set([start])
        nodes_explored = 0

        # Lista que prioriza o meio
        directions = [
            (1, 0, 'down'),
            (0, 1, 'right'),
            (-1, 0, 'up'),
            (0, -1, 'left')
        ]

        while queue:
            (curr_r, curr_c), path = queue.popleft()
            nodes_explored += 1

            if (curr_r, curr_c) == target:
                self.get_logger().info(f'--- RELATÓRIO ---')
                self.get_logger().info(f'Passos: {len(path)}')
                self.get_logger().info(f'Esforço: {nodes_explored}')
                return path

            for dr, dc, move_name in directions:
                next_r, next_c = curr_r + dr, curr_c + dc
                if 0 <= next_r < rows and 0 <= next_c < cols:
                    if grid[next_r][next_c] != 'b' and (next_r, next_c) not in visited:
                        visited.add((next_r, next_c))
                        queue.append(((next_r, next_c), path + [move_name]))
        return None

    def run(self):
    
        self.get_logger().info('Resetando o labirinto...')
        self.reset_game()
        time.sleep(0.5) 

        
        self.get_logger().info('Lendo o mapa...')
        map_res = self.get_map_data()
        
        rows, cols = map_res.occupancy_grid_shape
        flat_grid = map_res.occupancy_grid_flattened
        
        grid = []
        start_pos = None
        target_pos = None

        
        for r in range(rows):
            row_data = []
            for c in range(cols):
                val = flat_grid[r * cols + c]
                row_data.append(val)
                if val == 'r':
                    start_pos = (r, c)
                elif val == 't':
                    target_pos = (r, c)
            grid.append(row_data)

        
        if start_pos and target_pos:
            path = self.bfs(grid, start_pos, target_pos, rows, cols)
            if path:
                self.get_logger().info('Executando movimentos...')
                for step in path:
                    self.move_robot(step)
                    time.sleep(0.1)
            else:
                self.get_logger().error('Caminho não encontrado!')

def main(args=None):
    rclpy.init(args=args)
    resolver = MazeResolver()
    resolver.run()
    resolver.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cg_solution/src/resolver.cpp
#include <rclcpp/rclcpp.hpp>
#include <cg_interfaces/srv/move_cmd.hpp>
#include <cg_interfaces/srv/get_map.hpp>
#include <cg_interfaces/srv/reset.hpp>
#include <queue>
#include <vector>
#include <string>
#include <set>
#include <thread>
#include <iostream>

using namespace std;
using namespace rclcpp;
using namespace std::chrono_literals;

string grid_global[100][100];
int linhas_g, colunas_g;

class MazeResolver : public Node {
public:
    Client<cg_interfaces::srv::MoveCmd>::SharedPtr cl_move;
    Client<cg_interfaces::srv::GetMap>::SharedPtr cl_map;
    Client<cg_interfaces::srv::Reset>::SharedPtr cl_reset;

    MazeResolver() : Node("maze_resolver") {
        cl_move = create_client<cg_interfaces::srv::MoveCmd>("/move_command");
        cl_map = create_client<cg_interfaces::srv::GetMap>("/get_map");
        cl_reset = create_client<cg_interfaces::srv::Reset>("/reset");
        

        while (!cl_move->wait_for_service(1s)) cout << "aguardando move..." << endl;
        while (!cl_map->wait_for_service(1s)) cout << "aguardando map..." << endl;
        while (!cl_reset->wait_for_service(1s)) cout << "aguardando reset..." << endl;
    }

    void do_reset() {
        auto req = make_shared<cg_interfaces::srv::Reset::Request>();
        req->is_random = true;
        req->map_name = "";
        auto f = cl_reset->async_send_request(req);
        spin_until_future_complete(this->shared_from_this(), f);
    }

    void do_move(string d) {
        auto req = make_shared<cg_interfaces::srv::MoveCmd::Request>();
        req->direction = d;
        auto f = cl_move->async_send_request(req);
        spin_until_future_complete(this->shared_from_this(), f);
    }

    vector<string> bfs(pair<int,int> inicio, pair<int,int> alvo) {
        queue<pair<pair<int,int>, vector<string>>> q;
        set<pair<int,int>> visitados;
        
        vector<string> caminho_vazio;
        q.push({inicio, caminho_vazio});
        visitados.insert(inicio);
        
        int dr[] = {1, 0, -1, 0};
        int dc[] = {0, 1, 0, -1};
        string moves[] = {"down", "right", "up", "left"};

        while(!q.empty()) {
            auto item = q.front();
            q.pop();
            
            int r = item.first.first;
            int c = item.first.second;
            vector<string> path = item.second;

            if(r == alvo.first && c == alvo.second) {
                cout << "Caminho encontrado! Passos: " << path.size() << endl;
                return path;
            }

            for(int i=0; i<4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                
                if(nr >= 0 && nr < linhas_g && nc >= 0 && nc < colunas_g) {
                    if(grid_global[nr][nc] != "b" && visitados.count({nr, nc}) == 0) {
                        visitados.insert({nr, nc});
                        vector<string> new_path = path;
                        new_path.push_back(moves[i]);
                        q.push({{nr, nc}, new_path});
                    }
                }
            }
        }
        return {};
    }

    void start() {
        cout << "Resetando..." << endl;
        do_reset();
        this_thread::sleep_for(500ms);

        auto req_map = make_shared<cg_interfaces::srv::GetMap::Request>();
        auto fut_map = cl_map->async_send_request(req_map);
        spin_until_future_complete(this->shared_from_this(), fut_map);
        auto res = fut_map.get();

        linhas_g = res->occupancy_grid_shape[0];
        colunas_g = res->occupancy_grid_shape[1];
        auto flat = res->occupancy_grid_flattened;

        pair<int,int> start_pos;
        pair<int,int> target_pos;

        int k=0;
        for(int i=0; i<linhas_g; i++) {
            for(int j=0; j<colunas_g; j++) {
                string val = flat[k];
                grid_global[i][j] = val;
                if(val == "r") start_pos = {i, j};
                if(val == "t") target_pos = {i, j};
                k++;
            }
        }

        auto rota = bfs(start_pos, target_pos);
        
        if(rota.size() > 0) {
            cout << "Executando..." << endl;
            for(size_t i=0; i<rota.size(); i++) {
                do_move(rota[i]);
                this_thread::sleep_for(100ms);
            }
        } else {
            cout << "Erro: BFS nao achou nada." << endl;
        }
    }
};

int main(int argc, char **argv) {
    init(argc, argv);
    auto node = make_shared<MazeResolver>();
    node->start();
    shutdown();
    return 0;
}